Add zoneinfo data for System.Runtime.TimeZoneInfoTests

tqiu8 opened the pull request:
- added --enable-zoneinfo flags to WasmRunnerTemplate.sh as well as wasm/sample/Makefile
- added package reference to System.Runtime.TestData
- allows all TimeZoneInfo tests with `System.TimeZoneNotFoundException` to pass, ~100 tests

tqiu8 committed:
Add zoneinfo data for System.Runtime.TimeZoneInfoTests
- added --enable-zoneinfo flags to WasmRunnerTemplate.sh as well as wasm/sample/Makefile
- added package reference to System.Runtime.TestData

safern commented on eng/testing/WasmRunnerTemplate.sh:
```diff
@@ -10,7 +10,8 @@ XHARNESS_OUT="$EXECUTION_DIR/xharness-output"
     --js-file=runtime.js -v \
     --output-directory=$XHARNESS_OUT \
-    -- --enable-gc --run WasmTestRunner.dll ${@:2}
+    -- --enable-gc --enable-zoneinfo \
```
Should `--enable-zoneinfo` be enabled for all projects?

The way we construct this file is changing in: https://github.com/dotnet/runtime/pull/37752/files#diff-cb1e9a952e91b5f53a90e69eb6536127R14

I think this should be more a settting in the project itself, and then we can include this argument in the template, just as we pass down the XunitArguments. Basically once https://github.com/dotnet/runtime/pull/37752 is merged, we could add something like:

```xml
 <PropertyGroup Condition="'$(TargetOS)' == 'Browser'">
    <!-- We need to set this in order to get extensibility on xunit category traits and other arguments we pass down to xunit via MSBuild properties -->
    <BrowserRuntimeArguments>--enable-gc</BrowserRuntimeArguments>
    <BrowserRuntimeArguments Condition="'$(BrowserEnableZoneInfo)' == 'true'">$(BrowserRuntimeArguments) --enable-zoneinfo</BrowserRuntimeArguments>
    <RunScriptCommand>$HARNESS_RUNNER xharness wasm test --engine=$(JSEngine) --js-file=runtime.js -v --output-directory=$XHARNESS_OUT -- $(BrowserRuntimeArguments) --run WasmTestRunner.dll $(AssemblyName).dll</RunScriptCommand>
  </PropertyGroup>
```

Somewhere around here: https://github.com/dotnet/runtime/pull/37752/files#diff-9e21cdbaea18a370997e662e572afc6dR16

Then the project could just set:
```xml
<BrowserEnableZoneInfo>true</BrowserEnableZoneInfo>
```

safern replied:
I think merging this as is would break all test projects that don't have a reference to `System.Runtime.TestData` because it won't find the zoneinfo file when trying to read it:

```
read ("supportFiles/zoneinfo/mono-webassembly-zoneinfo-fs-smd.js.metadata", 'utf-8')
```

tqiu8 replied:
That makes sense. I will wait for that PR to be merged and then will implement those changes.

safern commented on src/mono/wasm/runtime-test.js:
```diff
@@ -189,6 +189,7 @@ var Module = {
 						{
 							var fullPath = directory != '/' ? directory + '/' + files[j] : files[j];
 							var content = new Uint8Array (read ("supportFiles/" + fullPath, 'binary'));
+							console.log(fullPath);
```
Do we need to log the fullPath or was this for testing purposes? I remember this making the output larger.

tqiu8 replied:
This was for debugging purposes. I will take that out.

safern commented on src/mono/netcore/sample/wasm/Makefile:
```diff
@@ -13,7 +13,7 @@ clean:
 
 run:
-	cd bin/$(CONFIG)/publish && ~/.jsvu/v8 --expose_wasm runtime.js -- --enable-gc --run WasmSample.dll
+	cd bin/$(CONFIG)/publish && ~/.jsvu/v8 --expose_wasm runtime.js -- --enable-gc --enable-zoneinfo --run WasmSample.dll 
```
This would also break the sample right? Does the sample have the `supportFiles`?

safern replied:
Oh nevermind, I see you're adding a package reference to it.

safern commented on src/mono/netcore/sample/wasm/WasmSample.csproj:
```diff
@@ -25,10 +25,14 @@
   <Target Name="BuildApp" DependsOnTargets="RebuildWasmAppBuilder;Build">
+  <ItemGroup>
+      <WasmFilesToIncludeInFileSystem Include="@(ContentWithTargetPath)" />
+      <WasmFilesToIncludeInFileSystem Include="@(ReferenceCopyLocalPaths)" Condition="'%(ReferenceCopyLocalPaths.BuildReference)' == 'true'" />
```
This was needed in the mobile targets just for ProjectReferences, do we have any in the `WasmSample`?

tqiu8 replied:
I don't think so. Let me take that out.

safern commented on src/mono/netcore/sample/wasm/WasmSample.csproj:
```diff
@@ -39,11 +43,24 @@
+  
+
+  <ItemGroup>
+    <PackageReference Include="xunit" Version="2.4.0" />
```
Why do we need a reference to `xunit`?

tqiu8 replied:
This was for writing tests in the Wasm sample app.

tqiu8 committed:
Remove --enable-zoneinfo from WasmRunnerTemplate.sh and remove extra references/logs

marek-safar commented on src/mono/netcore/sample/wasm/WasmSample.csproj:
```diff
@@ -39,11 +39,18 @@
   </Target>
 
+  <ItemGroup>
+    <PackageReference Include="System.Runtime.TestData" Version="$(SystemRuntimeTestDataVersion)" />
```
Including System.Runtime.TestData does not make much sense for WasmSample

tqiu8 replied:
Sounds good. I will just revert the sample to the original.

tqiu8 committed:
Remove extra imports and references from WasmSample

safern commented:
@tqiu8 the PR to add WASM test runs in CI was just merged so you should be able to rebase on top of master and address: https://github.com/dotnet/runtime/pull/38219#discussion_r443762714

Also, with this fix, are there any test failures left in `System.Runtime.Tests`?

tqiu8 commented:
@safern I was gonna go a different route and just have the timezone data included in the runtime pack instead of using the package reference.

safern commented:
Is that going to be the end-to-end scenario for customers?

tqiu8 commented:
@safern I think it's a temporary solution that will eventually be merged with katelyn's icu work. @lewing

tqiu8 committed:
Switch from loading timezonedata from runtime-assets to including dotnet.timezones.blat in runtime pack
Remove edits to WasmSample makefile
Remove package references to System.Runtime.TestData

safern commented on eng/Versions.props:
```diff
@@ -80,6 +80,7 @@
     <SystemPrivateRuntimeUnicodeDataVersion>5.0.0-beta.20319.2</SystemPrivateRuntimeUnicodeDataVersion>
     <SystemSecurityCryptographyX509CertificatesTestDataVersion>5.0.0-beta.20319.2</SystemSecurityCryptographyX509CertificatesTestDataVersion>
     <SystemWindowsExtensionsTestDataVersion>5.0.0-beta.20319.2</SystemWindowsExtensionsTestDataVersion>
+    <SystemRuntimeTestDataVersion>5.0.0-beta.20319.2</SystemRuntimeTestDataVersion>
```
I don't think we need this anymore, right?

tqiu8 replied:
yes I can remove that

tqiu8 committed:
Adding enable-zoneinfo the runscriptcommand

tqiu8 committed:
Merge branch 'master' into add-zoneinfo-data

safern commented:
Should we enable more test assemblies that were impacted by this?

lewing commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -779,6 +779,64 @@ var MonoSupportLib = {
 			return className.replace(/\//g, '.').replace(/`\d+/g, '');
 		},
+
+		wasm_load_timezone_data: function (file_name) {
```
this doesn't need to be timezone specific.  I think this should become something like mono_wasm_load_data where it takes the array as the first arg and possibly a second prefix arg and them move the read() call to the test runner before it calls this.

lewing commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -779,6 +779,64 @@ var MonoSupportLib = {
+			Module['FS_createPath']('/', 'usr', true, true);
+			FS.mkdir("share");
+			Module['FS_createPath']('/usr', 'share', true, true);
+			FS.mkdir("/usr/share/zoneinfo");
```
if the prefix is passed to the method you can just prepend it to m[0] and you can remove this extra code and it will be created in the folder generation

tqiu8 replied:
Hm, it doesn't seem to create /usr/share/zoneinfo during folder generation:
```
folders.forEach(folder => {
		FS.mkdir(`${prefix}/${folder}`)
		Module['FS_createPath'](prefix, folder, true, true);
		});
```

tqiu8 committed:
Incorporate PR feedback, make mono_wasm_load_data more general

tqiu8 committed:
Merge branch 'master' into add-zoneinfo-data

safern commented on eng/testing/tests.mobile.targets:
```diff
@@ -14,7 +14,7 @@
 
   <PropertyGroup Condition="'$(TargetOS)' == 'Browser'">
     <!-- We need to set this in order to get extensibility on xunit category traits and other arguments we pass down to xunit via MSBuild properties -->
-    <RunScriptCommand>$HARNESS_RUNNER xharness wasm test --engine=$(JSEngine) $(JSEngineArgs) --js-file=runtime.js -v --output-directory=$XHARNESS_OUT -- --run WasmTestRunner.dll $(AssemblyName).dll</RunScriptCommand>
```
Why are we removing `$(JSEngineArgs)`?

safern commented on eng/testing/tests.mobile.targets:
```diff
@@ -14,7 +14,7 @@
   <PropertyGroup Condition="'$(TargetOS)' == 'Browser'">
     <!-- We need to set this in order to get extensibility on xunit category traits and other arguments we pass down to xunit via MSBuild properties -->
-    <RunScriptCommand>$HARNESS_RUNNER xharness wasm test --engine=$(JSEngine) $(JSEngineArgs) --js-file=runtime.js -v --output-directory=$XHARNESS_OUT -- --run WasmTestRunner.dll $(AssemblyName).dll</RunScriptCommand>
+    <RunScriptCommand>$HARNESS_RUNNER xharness wasm test --engine=$(JSEngine) --js-file=runtime.js -v --output-directory=$XHARNESS_OUT -- ---enable-zoneinfo -run WasmTestRunner.dll $(AssemblyName).dll</RunScriptCommand>
```
It seems like you're accidentally stripping a `-` from `--run`.

tqiu8 replied:
argh sorry, this was probably a mistake from the merge.

tqiu8 committed:
Correct tests.mobile.targets

tqiu8 committed:
Pr feedback:
- Correct tests.mobile.targets
- Hardcode magic number
- Make folder depth variable

tqiu8 committed:
Merge branch 'master' into add-zoneinfo-data

lewing commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -779,6 +779,49 @@ var MonoSupportLib = {
+				var name = row[0];
+				var length = row[1];
+				var bytes = data.slice(0, length);
+				Module['FS_createDataFile'](`/usr/share/zoneinfo/${name}`, null, bytes, true, true);
```
```suggestion
				Module['FS_createDataFile'](`${prefix}/${name}`, null, bytes, true, true);
```

tqiu8 committed:
Replace prefix

Co-authored-by: Larry Ewing <lewing@microsoft.com>

lewing commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -779,6 +779,49 @@ var MonoSupportLib = {
+			manifest.filter(m => {
+				m = m[0].split('/')
+				if (m!= null) {
+					if (m.length === 3) folders.add(m.slice(0,2).join('/'));
```
This appears to limit the depth of the directory structure in the archive, can you write it so that it works with arbitrary depths?

lewing commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -779,6 +779,49 @@ var MonoSupportLib = {
 			return className.replace(/\//g, '.').replace(/`\d+/g, '');
 		},
+
+		mono_wasm_load_data: function (data, prefix, magic) {
```
I think we should hardcode the magic number here (but it looks reversed?).  If it isn't hardcoded and isn't defined it should skip the check entirely though

tqiu8 committed:
Merge branch 'master' into add-zoneinfo-data

akoeplinger commented on src/mono/wasm/Makefile:
```diff
@@ -112,6 +112,9 @@ clean:
 # Helper targets
 .PHONY: runtime
 
+timezone-data:
+	cp runtime/dotnet.timezones.blat $(BUILDS_BIN_DIR)
```
nit in case you touch this again: this should ideally use the filename as the target name instead so make's dependency tracking works (right now it always copies the file even if it's already there:

```make
$(BUILDS_BIN_DIR)/dotnet.timezones.blat: runtime/dotnet.timezones.blat
	cp runtime/dotnet.timezones.blat $@
```

(the `$@` is a special variable that means the target filename: https://www.gnu.org/software/make/manual/html_node/Automatic-Variables.html)

tqiu8 replied:
good to know. i can open another pr with this. thank you!
